File: packages/sdk-py/src/agentgate/message_context.py
"""MessageContext — wraps UserContext and provides reply/reply_stream methods."""
from __future__ import annotations
from typing import AsyncIterable, Any


class MessageContext:
    """
    Passed to the user's message handler. Exposes all UserContext fields
    and provides reply() / reply_stream() for sending responses.
    """

    def __init__(self, message_id: str, payload: dict, ws: Any) -> None:
        self._message_id = message_id
        self._ws = ws

        # UserContext fields — exposed as read-only attributes
        self.message: str = payload["message"]
        self.user: dict = payload["user"]
        self.role: str = payload["role"]
        self.permissions: list[str] = payload["permissions"]
        self.thread_id: str = payload["thread_id"]
        self.channel_id: str | None = payload.get("channel_id")
        self.is_superadmin: bool = payload.get("is_superadmin", False)
        self.sources: dict = payload.get("sources", {})

    async def reply(self, content: str) -> None:
        """Send a single complete response."""
        await self._ws.send({
            "type": "reply",
            "messageId": self._message_id,
            "content": content,
            "done": True,
        })

    async def reply_stream(self, generator: AsyncIterable[str]) -> None:
        """Stream response chunks, then send a done=True closer."""
        async for chunk in generator:
            await self._ws.send({
                "type": "reply",
                "messageId": self._message_id,
                "content": chunk,
                "done": False,
            })
        # Always send closer
        await self._ws.send({
            "type": "reply",
            "messageId": self._message_id,
            "content": "",
            "done": True,
        })

File: packages/sdk-ts/src/message-context.ts
import type { UserContext, Permission, AgentUser, AgentSources, ReplyFrame } from './types'
import type { WsClient } from './ws-client'

export class MessageContext {
  readonly message: string
  readonly user: AgentUser
  readonly role: string
  readonly permissions: Permission[]
  readonly thread_id: string
  readonly channel_id: string | null
  readonly is_superadmin: boolean
  readonly sources: AgentSources

  constructor(
    private readonly messageId: string,
    payload: UserContext,
    private readonly ws: Pick<WsClient, 'send'>
  ) {
    this.message = payload.message
    this.user = payload.user
    this.role = payload.role
    this.permissions = payload.permissions
    this.thread_id = payload.thread_id
    this.channel_id = payload.channel_id
    this.is_superadmin = payload.is_superadmin
    this.sources = payload.sources
  }

  async reply(content: string): Promise<void> {
    const frame: ReplyFrame = {
      type: 'reply',
      messageId: this.messageId,
      content,
      done: true,
    }
    this.ws.send(frame)
  }

  async replyStream(generator: AsyncIterable<string>): Promise<void> {
    for await (const chunk of generator) {
      const frame: ReplyFrame = {
        type: 'reply',
        messageId: this.messageId,
        content: chunk,
        done: false,
      }
      this.ws.send(frame)
    }
    const closer: ReplyFrame = {
      type: 'reply',
      messageId: this.messageId,
      content: '',
      done: true,
    }
    this.ws.send(closer)
  }
}
